### PR TITLE
#611, Use own last trade as basis for new center price

### PR DESCRIPTION
**Pull request by MarkoPaasila**
Related to #611. I'm still testing this but made this pr so I'm open to comments and suggestions.

I haven't managed to update the gui yet. I wouldn't mind if someone pointed out the necessary things to do.

### CODE REVIEW

File: dexbot/strategies/relative_orders.py [hunk 1]
```diff
@@ -58,10 +58,14 @@ def __init__(self, *args, **kwargs):
         if self.is_center_price_dynamic:
             self.center_price = None
             self.center_price_depth = self.worker.get('center_price_depth', 0)
+            self.cp_from_last_trade = self.worker.get('cp_from_last_trade')
```
Comment (bitfag) [inline dexbot/strategies/relative_orders.py]: Add default here.

File: dexbot/strategies/relative_orders.py [hunk 2]
```diff
@@ -197,6 +201,9 @@ def calculate_order_prices(self):
                 # Calculate with quote amount if given
                 center_price = self.get_market_center_price(quote_amount=self.center_price_depth)
 
+            if self.cp_from_last_trade:
+                center_price = self.get_own_last_trade()['price']
```
Comment (bitfag) [inline dexbot/strategies/relative_orders.py]: Add handling if `self.get_own_last_trade()` returned unusable result (like if it was no trades in account history).

File: dexbot/strategies/relative_orders.py [hunk 3]
```diff
@@ -583,3 +590,14 @@ def check_orders(self, *args, **kwargs):
...
+    def get_own_last_trade(self):
+        """ Returns dict with amounts and price of last trade """
+        trade = [x for x in self.account.history(limit=1, only_ops=['fill_order'])][0]['op'][1]
+        if trade['pays']['asset_id'] == self.market['base']['id']:  # Buy order
+            base = trade['fill_price']['base']['amount'] / 10 ** self.market['base']['precision']
+            quote = trade['fill_price']['quote']['amount'] / 10 ** self.market['quote']['precision']
+        else:  # Sell order
+            base = trade['fill_price']['quote']['amount'] / 10 ** self.market['base']['precision']
+            quote = trade['fill_price']['base']['amount'] / 10 ** self.market['quote']['precision']
+        return {'base': base, 'quote': quote, 'price': base / quote}
```
Comment (bitfag) [inline dexbot/strategies/relative_orders.py]: Add handling if `trade` was empty list

**Comment by MarkoPaasila** [comment]
Ok, error handling added. It still lacks the gui change. I'll  attempt.

**Comment by MarkoPaasila** [comment]
I have no idea how to add this checkbox to the configuration dialog. I found `class RelativeOrdersController(StrategyController):` which isn't used anywhere that I can find. I don't dare to touch it.

**Comment by MarkoPaasila** [comment]
I managed to add the checkbox to the widget, but it isn't connected to the logic yet. I would have copied how center_price_offset_input is connected, but it doesn't look like that's even connected - so I couldn't copy the implementation. Must do more research

File: dexbot/strategies/config_parts/relative_config.py [hunk 4]
```diff
@@ -54,6 +54,8 @@ def configure(cls, return_base_config=True):
             ConfigElement('center_price_depth', 'float', 0, 'Measurement depth',
                           'Cumulative quote amount from which depth center price will be measured',
                           (0.00000001, 1000000000, 8, '')),
+            ConfigElement('center_price_from_last_trade', bool, 1, 'Last trade price as new center price',
```
Comment (bitfag) [inline dexbot/strategies/config_parts/relative_config.py]: Change "1" to False

**Comment by bitfag** [comment]
GUI items logic is in dexbot/controllers/strategy_controller.py

**Comment by MarkoPaasila** [comment]
Still no success. I find it hard to debug as I don't know where the new config option should pop up but doesn't

File: dexbot/strategies/config_parts/relative_config.py [hunk 5]
```diff
@@ -54,6 +54,8 @@ def configure(cls, return_base_config=True):
             ConfigElement('center_price_depth', 'float', 0, 'Measurement depth',
                           'Cumulative quote amount from which depth center price will be measured',
                           (0.00000001, 1000000000, 8, '')),
+            ConfigElement('center_price_from_last_trade', bool, False, 'Last trade price as new center price',
```
Comment (joelvai) [inline dexbot/strategies/config_parts/relative_config.py]: Change `bool` -> `'bool'`

File: dexbot/strategies/relative_orders.py [hunk 6]
```diff
@@ -58,6 +58,7 @@ def __init__(self, *args, **kwargs):
         if self.is_center_price_dynamic:
             self.center_price = None
             self.center_price_depth = self.worker.get('center_price_depth', 0)
+            self.cp_from_last_trade = self.worker.get('cp_from_last_trade', False)
```
Comment (joelvai) [inline dexbot/strategies/relative_orders.py]: Change `self.worker.get('cp_from_last_trade', False)` -> `self.worker.get('center_price_from_last_trade', False)` to match the ConfigElement

File: dexbot/strategies/config_parts/relative_config.py (hunk 5 above)
Comment (MarkoPaasila) [reply dexbot/strategies/config_parts/relative_config.py]: Wow, how could I have noticed that

**Pull request merged**